### DISCUSSION TITLE
How to use ProxyAgent and CookieAgent at the same time?

### DISCUSSION THREAD

**Discussion by ickg5** [nodejs/undici#1957] [Q&A]
Hey folks, greetings 👋
I now need to use ProxyAgent and CookieAgent at the same time, I don't know how to write.
I don't see anything about this in the documentation.
Please help me.

My example code goes something like this:
```ts
import { ProxyAgent, request, setGlobalDispatcher } from "undici"
import { CookieAgent } from "http-cookie-agent/undici"
import { CookieJar } from "tough-cookie"


async function test() {
  const proxyAgent = new ProxyAgent("http://127.0.0.1:10810")
  setGlobalDispatcher(proxyAgent)
  const jar = new CookieJar()
  const agent = new CookieAgent({ cookies: { jar } })
  await request("https://test.com", {
    dispatcher: agent,
  })
}
```
Help would be greatly appreciated! 🙏

**Comment by mcollina** [comment]
I've opened a new issue: https://github.com/nodejs/undici/issues/1962

**Comment by dglozic** [comment]
While we are at it, is setting the global dispatcher the only way to do this? I would like to change some settings but I want those changed settings to be just for that one request, not for all subsequent requests. I see I can set a `dispatcher` in the options. Would I need to create a new Agent({}) and set it in options? Would I leak agents this way (do I need to dispose of them later)?

**Reply by mcollina** [reply]
> I see I can set a dispatcher in the options. Would I need to create a new Agent({}) and set it in options? 

I would recommend you reusing them whenever possible. Note that you can also call `agent.request()` and skip the option.

> Would I leak agents this way (do I need to dispose of them later)?

Ultimately no, but some sockets would linger around until they timeouts. I would recommend reusing them or destroying them.

**Comment by dglozic** [comment]
Perhaps I should open a new question :-).

**Comment by dglozic** [comment]
> I would recommend you reusing them whenever possible. 

OK, so what I can do is create a new Agent as a global variable, call it 'selfSignedDispatcher', set the connect to rejectAuthorized: false, and set it as a dispatcher for calls that use self-signed certificates. Thus I would only use either default global dispatcher or the one I created for reuse.
